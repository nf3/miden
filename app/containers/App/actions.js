/*
 * App Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

/** ----  Vault Vision changed code block ---- */
import {
  LOAD_REPOS,
  LOAD_REPOS_SUCCESS,
  LOAD_REPOS_ERROR,
  AUTH_SUCCESS,
  VALIDATE_AUTH_CALLBACK,
  VALIDATE_AUTH_ERROR,
  LOAD_USER,
  LOAD_USER_SUCCESS,
} from './constants';
/** ---- end block ----  */

/**
 * Load the repositories, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_REPOS
 */
export function loadRepos() {
  return {
    type: LOAD_REPOS,
  };
}

/**
 * Dispatched when the repositories are loaded by the request saga
 *
 * @param  {array} repos The repository data
 * @param  {string} username The current username
 *
 * @return {object}      An action object with a type of LOAD_REPOS_SUCCESS passing the repos
 */
export function reposLoaded(repos, username) {
  return {
    type: LOAD_REPOS_SUCCESS,
    repos,
    username,
  };
}

/**
 * Dispatched when loading the repositories fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of LOAD_REPOS_ERROR passing the error
 */
export function repoLoadingError(error) {
  return {
    type: LOAD_REPOS_ERROR,
    error,
  };
}

/**
 * Vault Vision added action for when a user gets a login token
 * Dispatched when the callback page is loaded to trigger the saga
 *
 *
 * @return {object}      An action object with the user data
 */
export function authSuccess(user) {
  return {
    type: AUTH_SUCCESS,
    user,
  };
}

/**
 * Vault Vision added action for when the auth token has bee processed and validated
 * Dispatched by the oidc/callback to initiate validation of the token returned by
 * the Vault Vision authentication platform
 *
 *
 */
export function validateAuthCallback() {
  return {
    type: VALIDATE_AUTH_CALLBACK,
  };
}

export function validateAuthError(error) {
  return {
    type: VALIDATE_AUTH_ERROR,
    error,
  };
}

/**
 * Vault Vision added action for when the application loads for the first time
 * and the application needs to check the local session state to see if a user
 * has already authenicated and is still in local session.
 *
 */
export function loadUser() {
  return {
    type: LOAD_USER,
  };
}

/**
 * Vault Vision added action for when a user gets returned by the loadUser
 * action.  This is the Success action to trigger the saving of the user
 * back into the global state
 *
 *
 * @return {object}      An action object with the user data
 */
export function userLoaded(user) {
  return {
    type: LOAD_USER_SUCCESS,
    user,
  };
}
