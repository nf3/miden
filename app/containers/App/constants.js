/*
 * AppConstants
 * Each action has a corresponding type, which the reducer knows and picks up on.
 * To avoid weird typos between the reducer and the actions, we save them as
 * constants here. We prefix them with 'yourproject/YourComponent' so we avoid
 * reducers accidentally picking up actions they shouldn't.
 *
 * Follow this format:
 * export const YOUR_ACTION_CONSTANT = 'yourproject/YourContainer/YOUR_ACTION_CONSTANT';
 */

export const LOAD_REPOS = 'boilerplate/App/LOAD_REPOS';
export const LOAD_REPOS_SUCCESS = 'boilerplate/App/LOAD_REPOS_SUCCESS';
export const LOAD_REPOS_ERROR = 'boilerplate/App/LOAD_REPOS_ERROR';

/** ----  Vault Vision added code block ---- */
// Added action for the login callback and the initial local session user load
export const AUTH_SUCCESS = 'boilerplate/App/AUTH_SUCCESS';
export const VALIDATE_AUTH_CALLBACK = 'boilerplate/App/VALIDATE_AUTH_CALLBACK';
export const VALIDATE_AUTH_ERROR = 'boilerplate/App/VALIDATE_AUTH_ERROR';

export const LOAD_USER = 'boilerplate/App/LOAD_USER';
export const LOAD_USER_SUCCESS = 'boilerplate/App/LOAD_USER_SUCCESS';
/** ---- end block ----  */
