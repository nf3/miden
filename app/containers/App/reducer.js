/*
 * AppReducer
 *
 * The reducer takes care of our data. Using actions, we can
 * update our application state. To add a new action,
 * add it to the switch statement in the reducer function
 *
 */

import produce from 'immer';

/** ----  Vault Vision changed code block ---- */
import {
  LOAD_REPOS_SUCCESS,
  LOAD_REPOS,
  LOAD_REPOS_ERROR,
  AUTH_SUCCESS,
  LOAD_USER_SUCCESS,
  VALIDATE_AUTH_CALLBACK,
  LOAD_USER,
  VALIDATE_AUTH_ERROR,
} from './constants';

// Added a user object and userLoading flag to global state
export const initialState = {
  loading: false,
  error: false,
  currentUser: false,
  user: null,
  userLoading: false,
  userData: {
    repositories: false,
  },
};

/* eslint-disable default-case, no-param-reassign */
const appReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case VALIDATE_AUTH_CALLBACK:
      case LOAD_USER:
        draft.userLoading = true;
        break;
      case AUTH_SUCCESS:
      case LOAD_USER_SUCCESS:
        console.log('action.user');
        console.log(action.user);
        draft.user = action.user;
        draft.userLoading = false;
        break;
      case LOAD_REPOS:
        draft.loading = true;
        draft.error = false;
        draft.userData.repositories = false;
        break;

      case LOAD_REPOS_SUCCESS:
        draft.userData.repositories = action.repos;
        draft.loading = false;
        draft.currentUser = action.username;
        break;

      case LOAD_REPOS_ERROR:
        draft.error = action.error;
        draft.loading = false;
        break;
      case VALIDATE_AUTH_ERROR:
        draft.error = action.error;
        draft.userLoading = false;
        console.log(action.error);
        break;
    }
  });

export default appReducer;
/** ---- end block ----  */
