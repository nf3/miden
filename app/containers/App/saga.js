/** ----  Vault Vision added code file ---- */
/**
 * Gets the auth state from any local session
 * Useful on App startup, especially on a refresh when a user is still logged in with a local session
 */

import { call, put, takeLatest } from 'redux-saga/effects';
import { LOAD_USER } from 'containers/App/constants';
import { validateAuthError, userLoaded } from 'containers/App/actions';

import { userManager } from '../../utils/userManager';

async function loadUserFromManager() {
  console.log('getting auth data');
  return userManager.getUser();
}

function* loadUser() {
  try {
    const user = yield call(loadUserFromManager);
    console.log(user);

    yield put(userLoaded(user));
  } catch (err) {
    yield put(validateAuthError(err));
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export default function* oidcLoadUser() {
  // Watches for LOAD_USER actions and calls loadData when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  // It will be cancelled automatically on component unmount
  yield takeLatest(LOAD_USER, loadUser);
}
