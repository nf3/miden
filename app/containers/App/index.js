/**
 *
 * App
 *
 * This component is the skeleton around the actual pages, and should only
 * contain code that should be seen on all pages. (e.g. navigation bar)
 */

import React, { useEffect } from 'react';
import { Helmet } from 'react-helmet';
import styled from 'styled-components';
import { Switch, Route } from 'react-router-dom';
import HomePage from 'containers/HomePage/Loadable';
import FeaturePage from 'containers/FeaturePage/Loadable';
import NotFoundPage from 'containers/NotFoundPage/Loadable';

/** ---- Vault Vision added imports for auth ---- */
import PropTypes from 'prop-types';
import { createStructuredSelector } from 'reselect';
import { connect, useDispatch } from 'react-redux';
import { compose } from 'redux';
import history from 'utils/history';

import { useInjectSaga } from 'utils/injectSaga';

import { makeSelectUser } from 'containers/App/selectors';

import DashboardPage from 'containers/Dashboard/Loadable';
import Profile from 'containers/Profile/Loadable';
import OidcCallback from 'containers/OidcCallback/Loadable';
import Auth from 'containers/Auth/Loadable';
/** ---- end block ----  */

import Header from 'components/Header';
import Footer from 'components/Footer';
import { loadUser } from './actions';
import saga from './saga';

/**
 * useAuth oidc-react hook gives us access to methods
 */

import GlobalStyle from '../../global-styles';
const key = 'app';

const AppWrapper = styled.div`
  max-width: calc(768px + 16px * 2);
  margin: 0 auto;
  display: flex;
  min-height: 100%;
  padding: 0 16px;
  flex-direction: column;
`;

export function App({ user }) {
  /** ----  Vault Vision added code block ---- */
  useInjectSaga({ key, saga });
  const dispatch = useDispatch();

  // This useEffect is to trigger an action that will reload a user from local session state
  // This is needed if a user is logged in and refreshes the page and the entire app reloads
  useEffect(() => {
    dispatch(loadUser());
  }, []);

  const isLoggedIn = () => {
    console.log(`user: ${user}`);
    if (user) {
      return true;
    }
    return false;
  };

  // This block sets up PrivateRoute feature that turns an route into an authentication required route
  const PrivateRoute = ({ component: Component, ...rest }) => (
    <Route
      {...rest}
      render={props =>
        isLoggedIn() === true ? <Component {...props} /> : history.push('/auth')
      }
    />
  );
  /** ---- end block ----  */

  return (
    <AppWrapper>
      <Helmet
        titleTemplate="%s - React.js Boilerplate"
        defaultTitle="React.js Boilerplate"
      >
        <meta name="description" content="A React.js Boilerplate application" />
      </Helmet>
      <Header />
      <Switch>
        <Route exact path="/" component={HomePage} />
        <Route path="/features" component={FeaturePage} />
        {/** ----  Vault Vision added code block ---- */}
        {/** The /auth route is what will initiate a new login flow, and redirect the user to Vault Vision login page */}
        <Route path="/auth" component={Auth} />
        {/** gets all request to /dashboard and renders the dashboardPage, and requires auth */}
        <PrivateRoute path="/dashboard" component={DashboardPage} />
        {/** gets all request to /profile and renders the profilePage, and requires auth */}
        <PrivateRoute path="/profile" component={Profile} />
        {/** The /oidc/callback route is what will receive the user returning from Vault Vision login page with a JWT */}
        <Route path="/oidc/callback" component={OidcCallback} />

        {/** ---- end block ---- */}
        <Route path="" component={NotFoundPage} />
      </Switch>
      <Footer />
      <GlobalStyle />
    </AppWrapper>
  );
}

/** ----  Vault Vision added code block ---- */
App.propTypes = {
  user: PropTypes.oneOfType([PropTypes.object, PropTypes.bool]),
};

const mapStateToProps = createStructuredSelector({
  user: makeSelectUser(),
});

const withConnect = connect(mapStateToProps);

export default compose(withConnect)(App);
/** ---- end block ----  */
