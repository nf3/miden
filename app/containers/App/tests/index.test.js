import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import { Provider } from 'react-redux';
import App from '../index';
import configureStore from '../../../configureStore';
import history from '../../../utils/history';

const renderer = new ShallowRenderer();

// Create redux store with history
const initialState = {};
const store = configureStore(initialState, history);

describe('<App />', () => {
  it('should render and match the snapshot', () => {
    renderer.render(
      <Provider store={store}>
        <App />
      </Provider>,
    );
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });
});
