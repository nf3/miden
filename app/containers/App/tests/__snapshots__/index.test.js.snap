// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should render and match the snapshot 1`] = `
<Context.Provider
  value={
    Object {
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "injectedReducers": Object {},
        "injectedSagas": Object {},
        "replaceReducer": [Function],
        "runSaga": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "subscription": Object {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <Memo(Connect(App)) />
</Context.Provider>
`;
